frontend/public/components/utils/resource-log.jsx

```diff
@@ -37,188 +32,202 @@ const LogControls = ({status, toggleStreaming, dropdown, onDownload}) => {
+  _toggleStreaming() {
+    const newStatus = this.state.status === 'streaming' ? 'paused' : 'streaming';
+    this.setState({
+      status: newStatus
     });
-    this._clearBuffer();
-    this._stream = stream(url, this._loadStarted, this._processData);
-    this._stream.promise
-      .then(() => {
-        // Resource is no longer running/generating new log content, so stop streaming
-        if (this.props.eof){
-          this._endStreaming();
-        } else {
-          // Request resolved, but resource is still in a non-terminated state, retry streaming
-          this._retry();
-        }
-      })
-      .catch((why) => { // Load failed/aborted
-        if (why !== 'abort') {
-          this._pushToBuffer(`Error: ${why}`);
-          this._retry();
-        }
-      });
   }
 
-  // Abort XMLHttpRequest
-  _endStreaming() {
-    this._stream && this._stream.abort();
+  // Updates log status
+  _updateStatus(newStatus) {
+    this.setState({status: newStatus});
   }
 
-  // Ends current stream and starts a new one.
-  _restartStreaming(){
-    this._endStreaming();
-    this._beginStreaming();
+  // Destroy websocket
+  _wsDestroy() {
+    this.ws && this.ws.destroy();
   }
 
-  _download () {
-    const blob = new Blob([this._buffer.lines().join('')], {type: 'text/plain;charset=utf-8'});
-    let filename = this.props.resourceName;
-    if (this.props.containerName) {
-      filename = `${filename}-${this.props.containerName}`;
-    }
-    saveAs(blob, `${filename}.log`);
+  // Initialize websocket connection and wire up handlers
+  _wsInit ({kind, namespace, resourceName, containerName, bufferSize}) {
+
+    const urlOpts = {
+      ns: namespace,
+      name: resourceName,
+      path: 'log',
+      queryParams: {
+        container: containerName || '',
+        follow: 'true',
+        tailLines: bufferSize
+      }
+    };
+
+    const watchURL = resourceURL(modelFor(kind), urlOpts);
+    const wsOpts = {
+      host: 'auto',
+      path: watchURL,
+      subprotocols: ['base64.binary.k8s.io']
+    };
+
+    this.ws = new WSFactory(watchURL, wsOpts)
+      .onclose(this._onClose.bind(this))
+      .onerror(this._onError.bind(this))
+      .onmessage(this._onMessage.bind(this))
+      .onopen(this._onOpen.bind(this));
   }
 
   render() {
+    const {dropdown, kind, bufferSize} = this.props;
+    const {error, lines, linesBehind, stale, status} = this.state;
+    const bufferFull = lines.length === bufferSize;
     return <React.Fragment>
+      {error && <p className="alert alert-danger">
+        <span className="pficon pficon-error-circle-o"></span>
+        An error occured while retrieving the requested logs.
+        <button className="btn btn-link" onClick={this._restartStream.bind(this)} >
+          Retry
+        </button>
+      </p>}
+      {stale && <p className="alert alert-info">
+        <span className="pficon pficon-info"></span>
+        The logs for this {kind.toLowerCase()} may be stale.
```

Comment:
Do we lowercase the kind elsewhere? I might just leave it as-is.

Reply:
My initial thought was just to use proper sentence case. I was also considering making this more generic because for pod logs, this message actually pertains to the currently selected container, but `Pod` would be used in this message instead of `Container`.